### DISCUSSION TITLE
Project specific globals in separate file?

### DISCUSSION THREAD

**Discussion by revision29** [PlummersSoftwareLLC/NightDriverStrip#544] [Ideas]
So, in my inability to observe things that are right in front of my face, I just noticed that in effects.cpp it checks for the existence of `custom_effects.h` and loads effects from it if present. This bypasses all of the other effects in the following elif conditions. I have been cluttering the effects.cpp file with effects related to the several projects in progress. "Finding" this bit of code will help me clean up my code some.

Then, it occurred to me, should "we" do the same for the globals.h file? Something like `custom_globals.h`. This would help separate the "stock" configurations from the one's defined by the user. This would greatly reduce my fear that a sync with main would nuke the several custom sections I have placed in the globals.h file. It's super easy to implement (I just did while typing this post). Just thought I would run the idea here instead of using a PR to vet an idea as I normally do.

**Comment by robertlipe** [comment]
On Sat, Nov 25, 2023, 8:02 PM Joe Schneider ***@***.***>
wrote:

> So, in my inability to observe things that are right in front of my face,
> I just noticed that in effects.cpp it checks for the existence of
> custom_effects.h and loads effects from it if present. This bypasses all
> of the other effects in the following elif conditions. I have been
> cluttering the effects.cpp file with effects related to the several
> projects in progress. "Finding" this bit of code will help me clean up my
> code some.
>
I created that and it was exactly my intention. Glad it helped!

> Then, it occurred to me, should "we" do the same for the globals.}}9⁰⁰h
> file? Something like custom_globals.h. This would help separate
>
I'd thought about it, but IMO adding more to that file is earely a good
idea.  Almost every time I started there, I recognized that adding a define
to the top level build (and outside the source tree) scratched my itch
perfectly.

That's probably not true for every combination.

>  placed in the globals.h file. It's super easy to implement (I just did
> while typing this post).
>
I'd gladly review or even create a new one.

I got hung up at one for boards, one for defaults,  one for effects, and so
on.  Let's not let the perfect be the enemy of the good.

> Just thought I would run the idea here instead of using a PR to vet an
> idea as I normally do.
>
That's probably not a bad idea anyway... Get consensus before you commit to
slinging bits unless you're committed to keeping those bits in your own
trees forever.

RJL

**Comment by rbergen** [comment]
This is _exactly_ what I mentioned in [this comment](https://github.com/PlummersSoftwareLLC/NightDriverStrip/discussions/538#discussioncomment-7654966) on #538. 

As that comment states, I don't think there's any harm in replacing `#if DEMO` in globals.h with 

```cpp
#if __has_include ("custom_globals.h")

      #include "custom_globals.h"

#elif DEMO
```

It doesn't address the fact that there are defines before and after the configuration blocks, but it does give more freedom to "play" with defines without impacting the main globals.h. Also, it provides the possibility without _any_ obligation to use it.

**Reply by revision29** [reply]
As you know I get lost in the comments. The downside that a person has to re-include a bunch of stuff, but that is true of the custom_effects.h file as well. As a matter of not confusing people new to the project, I would still keep demo in the main file, but obviously change to an elif statement. If it okay, I'd be glad to create a new PR for it for all of 4 lines of code.

If only we could do the same with platformio.ini....

**Reply by revision29** [reply]
After I typed the last line, I did a quick web search and found, that yes, a person can include external platformio config files from platformio.ini. [Documentation Here](https://docs.platformio.org/en/latest/projectconf/sections/platformio/options/generic/extra_configs.html)

**Reply by rbergen** [reply]
> I would still keep demo in the main file, but obviously change to an elif statement

I believe that's what I'm proposing. :)

> If it okay, I'd be glad to create a new PR for it for all of 4 lines of code.

I say: go ahead!

**Reply by rbergen** [reply]
> After I typed the last line, I did a quick web search and found, that yes, a person can include external platformio config files from platformio.ini.

Yes, that is possible. What I'm not sure of is if it's possible to include a file in platformio.ini _on the condition that the file exists_. Also, I think the whole "optional include" concept is trickier with platformio.ini, because the file follows a more rigid structure than the C++ files do.

**Reply by revision29** [reply]
In the platformio.ini, I don't think it is possible to add a conditional. But, now that I know I can, I will manually add the include according to the documentation. I'd rather "patch" two lines of code if my downstream file gets overwritten by a sync with main than have to go and copy and paste the 6-8 device/environment specifications I have in place. I was just happy to see pio supports the external file and thought I would share that.

I will make a PR for this once the other one is dealt with. After that I have another simple PR (6-8 lines) and plan on being done with PRs for a while.

**Reply by rbergen** [reply]
Ok. When you open the PR for the introduction of custom_globals.h, I think we should also add a line to .gitignore: `**/custom_*.h`

That will prevent people from accidentally including whatever custom headers they've decided to put together, to their commits, and hence their PRs.
